pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -2017,6 +2027,25 @@ private void handleLastMessageIdFromCompactedLedger(PersistentTopic persistentTo
         });
     }
 
+    private int calculateTheLastBatchIndexInBatch(MessageMetadata metadata, ByteBuf payload) throws IOException {
+        int batchSize = metadata.getNumMessagesInBatch();
+        if (batchSize <= 1){
+            return -1;
+        }
+        SingleMessageMetadata singleMessageMetadata = new SingleMessageMetadata();
+        int lastBatchIndexInBatch = -1;
+        for (int i = 0; i < batchSize; i++){
+            ByteBuf singleMessagePayload =
+                    Commands.deSerializeSingleMessageInBatch(payload, singleMessageMetadata, i, batchSize);
+            singleMessagePayload.release();
+            if (singleMessageMetadata.isCompactedOut()){
+                continue;
+            }
+            lastBatchIndexInBatch = i;
+        }
+        return lastBatchIndexInBatch;
```

Comment:
```suggestion
        for (int i = batchSize - 1; i >= 0; i--) {
            ByteBuf singleMessagePayload =
                    Commands.deSerializeSingleMessageInBatch(payload, singleMessageMetadata, i, batchSize);
            singleMessagePayload.release();
            if (!singleMessageMetadata.isCompactedOut()) {
                return i;
            }
        }
        return -1;
```

Use a reversed iteration to avoid unnecessary iterations because the most of time

Reply:
This is a good suggestion, but batch messages can only be parsed from front to back (^_^)

Reply:
Oh I see. It's limited by the implementation of `Commands#deSerializeSingleMessageInBatch`. Since the GetLastMessageId should not be a hot path, it's okay.

Reply:
Maybe we can consider adding a new field in message metadata to indicate the last valid message index, I'm afraid checking single metadata will cost too many resources.